pkg/db/connect.go

```diff
@@ -60,9 +63,21 @@ func ConnectDBPool(ctx context.Context, p params.Database) (*pgxpool.Pool, error
 	config.MaxConnLifetime = p.ConnectionMaxLifetime
 
 	pool, err := pgxpool.ConnectConfig(ctx, config)
-	if err != nil {
-		return nil, fmt.Errorf("could not open DB: %w", err)
+	for i := 0; i < DefaultMaxDatabaseRetries; i++ {
+		if err != nil {
+			if i != DefaultMaxDatabaseRetries-1 {
+				waitTime := math.Pow(RetryBase, float64(i))
+				log.Info(fmt.Sprintf("could not open DB: trying again in %ds", int(waitTime)))
+				time.Sleep(time.Duration(waitTime) * time.Second)
+				pool, err = pgxpool.ConnectConfig(ctx, config)
+
+				continue
+			} else {
+				return nil, fmt.Errorf("could not open DB: %w", err)
+			}
+		}
 	}
```

Comment:
This block is a bit hard to understand, possibly because you try to connect in one iteration but test the result in the next.  That also manifests in having to call `ConnectConfig` on two lines of code.  Also it means that you have the strange return just before the loop exits on l.76.

You might write the loop more directly:
```suggestion
    	for i := 0; i < DefaultMaxDatabaseRetries; i++ {
    	   	pool, err := pgxpool.ConnectConfig(ctx, config)
    	   	if err == nil {
    	   	    	return pool, nil
    	    	}
    	    	if i < DefaultMaxDatabaseRetries-1 {
    	    	    	wait := time.Duration(math.Pow(RetryBase, float64(i))) * time.Second
    	    	    	log.Info("open DB: %s.  Trying again in %s", wait)
    			time.Sleep(wait)
    		}
    	}
    	return nil, fmt.Errorf("open DB: %w", err)
```

And I would probably change the wait to
```go
wait := firstWait * pow(waitGrowth, i)
```
for some `const firstWait time.Duration` and `const waitGrowth float64`.  I like to retry aggressively at first (wait a WHOLE SECOND?), so e.g.
```go
const (
    	firstWait = 50 * time.MilliSecond
    	waitGrowth = 1.2
    	numRetries = /* whatever makes it fail within 3 seconds */
)
```
*Now* we can go all the way, and replace `numRetries` with a `totalWait` parameter so that we can say "just repeat this, with exponential backoff, I'm willing to wait up to `totalWait`.

Alternatively, what I like to do is have a "backoff iterator" that tells me whether to go again and how long to wait before I do.  E.g. I found this "retry" package has https://pkg.go.dev/gopkg.in/retry.v1.

Reply:
Thank you for the comments Ariel! I really appreciate the advice and the chance to improve my programming. If it's okay, I'll commit these changes and then ask for another review!